File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        return checkBST(root, Integer.MIN_VALUE,Integer.MAX_VALUE);
        
    }
    private boolean checkBST(TreeNode n, int min, int max ) {
        if (n == null) return true; 
        if (n.val <= min|| n.val >= max)return false;
        
        return checkBST(n.left, min, n.val)&& 
        checkBST(n.right,n.val, max);

        
    
    
        
    }

}

    

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, maxSoFar: int):
            if not node: return 0
            count = 1 if node.val>= maxSoFar else 0
            maxSoFar  = max(maxSoFar, node.val)
            count += dfs(node.right, maxSoFar)
            count += dfs(node.left, maxSoFar)
            return count
        return dfs(root, float('-inf'))

        